Source/Classes/AnimatedImages/PINCachedAnimatedImage.m

```diff
@@ -0,0 +1,426 @@
+//
+//  PINCachedAnimatedImage.m
+//  PINRemoteImage
+//
+//  Created by Garrett Moon on 9/17/17.
+//  Copyright © 2017 Pinterest. All rights reserved.
+//
+
+#import "PINCachedAnimatedImage.h"
+
+#import "PINRemoteLock.h"
+#import "PINGIFAnimatedImage.h"
+#if PIN_WEBP
+#import "PINWebPAnimatedImage.h"
+#endif
+
+#import <PINOperation/PINOperationQueue.h>
+#import "NSData+ImageDetectors.h"
+
+static const NSUInteger kFramesToRenderForLargeFrames = 4;
+static const NSUInteger kFramesToRenderMinimum = 2;
+
+static const CFTimeInterval kSecondsAfterMemWarningToMinimumCache = 1;
+static const CFTimeInterval kSecondsAfterMemWarningToLargeCache = 5;
+static const CFTimeInterval kSecondsAfterMemWarningToAllCache = 10;
+#if PIN_TARGET_IOS
+static const CFTimeInterval kSecondsBetweenMemoryWarnings = 15;
+#endif
+
+@interface PINCachedAnimatedImage () <PINCachedAnimatedFrameProvider>
+{
+    // Since _animatedImage is set on init it is thread-safe
+    id <PINAnimatedImage> _animatedImage;
+    
+    PINImage *_coverImage;
+    PINAnimatedImageInfoReady _coverImageReadyCallback;
+    dispatch_block_t _playbackReadyCallback;
+    NSMutableDictionary *_frameCache;
+    NSInteger _playbackReady; // Number of frames to cache until playback is ready
+    PINOperationQueue *_operationQueue;
+    dispatch_queue_t _cachingQueue;
+    
+    NSUInteger _playhead;
+    BOOL _notifyOnReady;
+    NSMutableIndexSet *_cachedOrCachingFrames;
+    PINRemoteLock *_lock;
+}
+
+@property (atomic, strong) NSDate *lastMemoryWarning;
+@property (atomic, assign) BOOL weAreTheProblem;
```

Comment:
I dig the name, but might be worth a comment explaining the meaning of this property ;)